lib/Adapter/PgsqlAdapter.php

```diff
@@ -143,16 +148,43 @@ public function set_encoding(string $charset): void
      *
      * @return array<string>
      */
-    public function determineColumnsThatNeedEscaping(array $columns): array
+    private function determineColumnsThatNeedEscaping(array $columns): array
     {
-        $ret = [];
+        $hash = implode(',', $columns);
+        $ret = $this->columnLookup[$hash] ?? null;
 
-        foreach ($columns as $column) {
-            if ($column !== strtolower($column)) {
-                $ret[] = $column;
+        if (null === $ret) {
+            $ret = [];
+
+            foreach ($columns as $column) {
+                if ($column !== strtolower($column)) {
+                    $ret[] = $column;
+                }
             }
+
+            $this->columnLookup[$hash] = $ret;
         }
 
         return $ret;
     }
+
+    /**
+     * @see Connection::escapeColumns()
+     *
+     * @param string        $expression The where clause to be escaped
+     * @param array<string> $columns    The columns of the table
+     */
+    public function escapeColumns(string $expression, array $columns): string
```

Comment:
Better, but still a bit complex, isn't it? I would think you could do it all in one method, something like:

```php
public function escapeColumns(string $expression, array $columns): string 
{
  $cachedNames = [];
   foreach ($columns as $column) {
      if ($column !== strtolower($column)) {
          $cachedNames[$column] ??= $this->quote_name($column);
          $expression = str_replace($column, $cachedNames($column), $expression);
      }
  }

   return $expression;
}
```

I didn't test it (I think I lost and implode in there, somewhere), and I'm sure you can think of something better, but maybe it'll give you a few ideas.

Reply:
New implementation is a bit shorter and more performant. ```escapeColumns()``` is called several times over with the exact same ```$columns``` passed in. Converting that ```$columns``` into a lookup hash prevents having to iterate over and over it again.

Reply:
Then you can use a static:

```
   static $cachedNames = [];
   foreach ($columns as $column) {
      if ($column !== strtolower($column)) {
          $cachedNames[$column] ??= $this->quote_name($column);
          $expression = str_replace($column, $cachedNames($column), $expression);
      }
  }

   return $expression;
```

Reply:
This is not as performant as the original implementation. The ```foreach ($columns as $column) {``` will be run over and over again for all the columns of a table. Most of them if you're following Postgres best practices will be lower cased.

Reply:
Yeah, I'll take care of it, thanks.